src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -0,0 +1,277 @@
+                Document document = editor.getDocument();
+                CharSequence documentChars = document.getCharsSequence();
+                formatTextRange = ContainerUtil.find(
+                        selectionTextRanges,
+                        selectionTextRange -> {
+                            int startOffset = selectionTextRange.getStartOffset();
+                            int endOffset = selectionTextRange.getEndOffset();
+
+                            // Remove leading newlines from the range
+                            while ((startOffset < endOffset) && (documentChars.charAt(startOffset) == '\n'))
+                                startOffset++;
+                            // Remove trailing whitespace up until newlines (inclusive) from the range
+                            boolean foundNewline = false;
+                            while (endOffset > startOffset) {
+                                char previousEndChar = documentChars.charAt(endOffset - 1);
+                                if (!Character.isWhitespace(previousEndChar)) {
+                                    break;
+                                } else if (previousEndChar == '\n') {
+                                    endOffset--;
+                                    foundNewline = true;
+                                } else if (!foundNewline) {
+                                    endOffset--;
+                                } else {
+                                    break;
+                                }
+                            }
+
+                            // See if this is a selection of complete lines
+                            int startLineNumber = document.getLineNumber(startOffset);
+                            int startLineStartOffset = document.getLineStartOffset(startLineNumber);
+                            if (startLineStartOffset == startOffset) {
+                                int endLineNumber = document.getLineNumber(endOffset);
+                                int endLineEndOffset = document.getLineEndOffset(endLineNumber);
+                                if ((endLineEndOffset == endOffset) || (endLineEndOffset == (endOffset + 1))) {
+                                    // Make sure that it ends with the terminator that was just typed
+                                    String selectionRangeText = StringUtil.trimTrailing(documentChars.subSequence(startOffset, endLineEndOffset).toString());
+                                    return ((startOffset + selectionRangeText.length()) == offset) && selectionRangeText.endsWith(String.valueOf(statementTerminatorChar));
+                                }
+                            }
+                            return false;
+                        }
+                );
+            }
+        }
+
+        // If appropriate, find the enclosing code block to format
+        else if (formattingScope == FormattingScope.CODE_BLOCK) {
```

Comment:
Again, it's valid to configure format-on-statement terminator to be for a more broad scope including the surrounding code block or file if that works better (or at all) for the associated language server. The default, of course, is statement scope.